### PR TITLE
Do not convert throwaway input from gcc tool.

### LINKED ISSUE #4011: UnicodeDecodeError
Why i get error if type help command?
Python 3.8.10
Scons 4.2.0
Безымянный

### PR DESCRIPTION
**Pull request by mwichmann**
In case we are in a locale where strings don't convert easily, only try to decode the first line of `gcc -version` output, as it's the only one we care about (and it's usually not translated). For the loop which throws away the remaining lines, don't try to convert them.

Hacked together an end-to-end test for this with a dummy gcc which reads known "bad" data from a file and sends it out on stdout.

Fixes #4011.

Signed-off-by: Mats Wichmann <mats@linux.com>

## Contributor Checklist:

* [ ] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [ ] I have updated `CHANGES.txt` (and read the `README.rst`)
* [ ] I have updated the appropriate documentation

### CODE REVIEW

File: CHANGES.txt [hunk 1]
```diff
@@ -19,6 +19,9 @@ RELEASE  VERSION/DATE TO BE FILLED IN LATER
       from 3.10 was not expected by SubstTests.py and test/Subst/Syntax.py
     - EmitterProxy rich comparison set is completed (checker warning).
       Added __le__, __gt__, __ge__.
+    - gcc tool when reading throwaway input does not need to call to_str.
```
Comment (bdbaddog) [inline CHANGES.txt]: Fix gcc/g++ tool failing when non utf-8 text is output from "gcc --version". This is run when gcc/g++ tools are initialized to determine which version of gcc is being run.

**Comment by bdbaddog** [comment]
For a test, can we use the .txt file the user provided and make a moc gcc which just outputs that text?

**Comment by mwichmann** [comment]
> For a test, can we use the .txt file the user provided and make a moc gcc which just outputs that text?

I'll think some more about that, I couldn't make it work before.

**Comment by bdbaddog** [comment]
Would it be welcome if I took a wack at a test case for this?

**Comment by mwichmann** [comment]
you're joking, right? :)  someone object to someone else trying to produce a test?  to "do it right" you need it to look like you're reading from the returned stdout from the subprocess call, which is an `<class '_io.BufferedReader'>`.  Because the routine in gcc.py uses `_subproc`, it's not completely clear to me how to do that "cleanly". Probably overthinking it.

**Comment by bdbaddog** [comment]
> you're joking, right? :) someone object to someone else trying to produce a test? to "do it right" you need it to look like you're reading from the returned stdout from the subprocess call, which is an `<class '_io.BufferedReader'>`. Because the routine in gcc.py uses `_subproc`, it's not completely clear to me how to do that "cleanly". Probably overthinking it.

I was thinking just make CC="python my_gcc.py". which then reads the example output in binary mode and writes in binary mode to stdout?

**Comment by mwichmann** [comment]
By the way, I have the "original" text that causes the problem, which doesn't look like what the OP attached to the issue - so I think there's a particular set of text, and then there's a Windows code page involved to further break stuff. The original is this:
```C
msgid ""
"This is free software; see the source for copying conditions.  There is NO\n"
"warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n"
"\n"
msgstr ""
"Это свободно распространяемое программное обеспечение. Условия копирования\n"
"приведены в исходных текстах. Без гарантии каких-либо качеств, включая \n"
"коммерческую ценность и применимость для каких-либо целей.\n"
"\n"
```

**Comment by mwichmann** [comment]
worth a try.

File: CHANGES.txt (hunk 1 above)
Comment (mwichmann) [reply CHANGES.txt]: okay, addressed that one.

**Pull request merged**